### PR TITLE
Delay releasing joining domains

### PR DESCRIPTION
**Pull request by kayceesrk**
Following the discussions in #11196, this PR partially reverts the changes in #11196. 

We now release the joining domains only after the `domain_terminate` function in the runtime has been completed (as was the case before #11196). This ensures that the terminating domain releases its resources such as heaps, ephemerons and finalisers before waking up joining domains. This obviates the need for PR #11208. We also additionally block signals on the terminating domain so that C signal handlers do not get to run on this thread.

The PR retains the simplifications on the OCaml side for `Domain.join` introduced by #11196. We use condition variables for parking joining domains, which avoids hand-written spin-wait loops.

### CODE REVIEW

**Comment by gasche** [comment]
> The PR retains the simplifications on the OCaml side for Domain.join introduced by [#11196](https://github.com/ocaml/ocaml/pull/11196). We use condition variables for parking joining domains, which avoids hand-written spin-wait loops.

Ah, thanks! I wanted to suggest this but did not get the time to.

**Comment by gasche** [comment]
Here is how I understand the spawn/join protocol proposed by this PR:

- startup: the OCaml implementation of Domain.spawn creates a "domain mutex" and a condition variable over it, to store in the `Domain.t` value; at this point the domain mutex is *not* locked and it will never get locked during startup
- startup: the OCaml code calls `Raw.spawn` which goes to the C implementation of `spawn` in the runtime, and passes it an OCaml callback to run as domain code *and* the domain mutex
- startup: the domain mutex is stored, in the C side (in the `domain_startup_params`), registered as a generational GC root to not be collected (even if the Domain.t value goes dead) while the domain is still running
- join: "joining" a domain is implemented purely on the OCaml side: you lock its domain mutex, and then wait on its condition variable (which un-locks the domain mutex)
- Termination: when the domain callback on the OCaml side is done computing its value, it locks the domain mutex and broadcasts the condition variable; it returns with the domain mutex still locked. In particular, at this point people waiting on the condition variable cannot start executing yet, as they need to take the domain mutex first.
- Then control goes back from the domain callback to the C code (the "termination" part of `domain_thread_func` passed to pthread). The C code gets the domain mutex from `domain_startup_params`, terminates the domain (all C-side cleanup logic), unlocks the termination mutex at this point, and then un-registers it as a GC root.

We end up with more C-side logic than in #11196, but this is simpler than the pre-#11196 implementation because we don't need spinning. (Spinning was required because we joined only by taking the domain lock, which was initially locked by the spawned domain and held during its complete duration, but this introduced a risk of race with the parent domain trying to join the child before it was done taking that lock. We could also have used a semaphore or condition variable between the child and the parent only, but the current design is arguably natural/idiomatic.)

**Comment by gasche** [comment]
Two things remain a bit complex with this approach:
1. the domain mutex is locked on the OCaml side inside the domain callback, but unlocked on the C side after the callback returns. This sort of non-symmetry is always a bit surprising at first.
2. the domain mutex is "stored" both on the OCaml side and the C side

## Idea to simplify (2)

A potential approach would be to not store the domain mutex on the C side, but instead have the `caml_domain_spawn` callback *return* the locked mutex (currently it has type `unit -> unit`). We would not need any specific root-registration then (having to return the mutex from the callback keeps it alive), it clarifies that the mutex is only used after the callback has terminated. This approach would not work if the callback could fail with an exception, but the runtime code of `domain_thread_func` already assumes that its callback never fails with a non-fatal error.

## Idea to simplify (1)

Do we really want to implement `Domain.join` in OCaml? If we were willing to implement it in the runtime, we could go back to the pre-#11196 approach of holding a lock, by having the child take the lock during the domain startup routine (before the status flips to Dom_started; there is already some spinning on the C side which could be reused here) and storing it in the domain-internal data.

Then joining another domain needs to get the internal domain (non-unique) identifier (those would be stored in the Domain.t value I guess), and call the runtime to (release its own domain runtime lock and) take the termination lock of the other domain then releasing it immediately.

This gives a pure-C implementation that would be simpler than the pre-#11196 implementation (which mixes C and OCaml logic, and needs spinning on the OCaml side as well). I'm not sure whether it would really be simpler than the current implementation. I suspect that it may still be simpler overall: we add an extra "termination lock" to the domain-internal state (there is already a lock there for backup thread stuff, but I don't think they are doing the same thing), but we remove some conceptual complexity at the interface between the OCaml and C code.

**Comment by gasche** [comment]
(Note: I made two potential suggestions above. The first is relatively simple and, if it turns out to make sense, could be pushed as a small simplification to the present PR. The second is a fairly different approach outside the scope of the current PR. It would make sense to delay the second question to later completely, and in any case merge the present PR which fixes an issue with the current state.)

**Comment by kayceesrk** [comment]
> We would not need any specific root-registration then (having to return the mutex from the callback keeps it alive)

I am not sure if this is true. The usual way to keep values alive from C is to use `CAMLlocalx`. But observe that the mutex is released after `domain_terminate` is finished. Since the thread no longer participates in the GC, the fact that it is registered using `CAMLlocalx` is not enough. 

We’re you thinking about something else @gasche?

**Comment by gasche** [comment]
I hadn't thought of this subtlety indeed. We are in a somewhat odd situation where we are running code on a thread that has disabled its own domain runtime, so domain-local mechanisms to handle OCaml values are not active anymore, but "global" mechanisms may still be used.

At this point we don't really need to keep the mutex value alive (we use the underlying system mutex in the C runtime and will never use the OCaml value again), but we want to prevent the finaliser on the OCaml mutex from running -- as nicely explained in the comments, thanks! This is a bit subtle (the pure-C solution avoids this contorsion as well) but it works, and for this using a global root seems more robust indeed.

File: runtime/domain.c [hunk 1]
```diff
@@ -1732,6 +1754,12 @@ static void domain_terminate (void)
   caml_plat_signal(&domain_self->domain_cond);
   caml_plat_unlock(&domain_self->domain_lock);
 
+#ifndef _WIN32
```
Comment (sadiqj) [inline runtime/domain.c]: I think this should be `#ifdef POSIX_SIGNALS`.

Also, are we doing this too late? A signal can still arrive after we've reached a stage where there's no more polling points on this domain. The issue is we'll set `young_limit` but then never check it, leading to a delayed signal (eventually another domain will probably do a GC or terminate).

Comment (kayceesrk) [reply runtime/domain.c]: > I think this should be #ifdef POSIX_SIGNALS.

I was going by the other uses of `pthread_sigmask` in `domain.c` which are protected by `#ifndef _WIN32`. If this should be changes, perhaps the other ones too. CC @dra27. 

> Also, are we doing this too late? A signal can still arrive after we've reached a stage where there's no more polling points on this domain. The issue is we'll set young_limit but then never check it, leading to a delayed signal (eventually another domain will probably do a GC or terminate).

If we move this to be the first thing that we do in `domain_terminate`, is that any better? We don't poll for pending signals in `domain_terminate` IINM. No OCaml callbacks on this domain can run once we've reached `domain_terminate`.

Comment (dra27) [reply runtime/domain.c]: `POSIX_SIGNALS` is the better one - I should have used that when I did the others!

Comment (kayceesrk) [reply runtime/domain.c]: Thanks. Let me use `POSIX_SIGNALS` for this one. Will you make a PR for the existing ones on trunk @dra27?

File: runtime/domain.c [hunk 2]
```diff
@@ -1084,16 +1089,30 @@ static void* domain_thread_func(void* v)
...
     caml_callback(ml_values->callback, Val_unit);
     domain_terminate();
+
+    term_mutex = Mutex_val(ml_values->term_mutex);
+    /* This domain currently holds the [term_mutex] lock, and has singaled
+       all the waiting domains to be woken up. We unlock the [term_mutex]
+       to release the joining domains. The unlock is done after
+       [domain_terminate] to ensure that this domain has released all of its
+       runtime state. */
+    caml_mutex_unlock(term_mutex);
```
Comment (xavierleroy) [inline runtime/domain.c]: You could get rid of `term_mutex` (again) and just write
```
caml_mutex_unlock(Mutex_val(ml_values->term_mutex));
```

File: runtime/domain.c [hunk 3]
```diff
@@ -1084,16 +1089,30 @@ static void* domain_thread_func(void* v)
     caml_domain_set_name("Domain");
     caml_callback(ml_values->callback, Val_unit);
     domain_terminate();
+
+    term_mutex = Mutex_val(ml_values->term_mutex);
+    /* This domain currently holds the [term_mutex] lock, and has singaled
```
Comment (xavierleroy) [inline runtime/domain.c]:
```suggestion
    /* This domain currently holds the [term_mutex] lock, and has signaled
```

**Review by xavierleroy — APPROVED** [review]
Looks good to me, thanks!  The locking / unlocking strategy for the termination mutex is clever.  Minor suggestions below.

File: runtime/domain.c (hunk 1 above)
Comment (dra27) [reply runtime/domain.c]: See #11219

File: runtime/domain.c (hunk 2 above)
Comment (kayceesrk) [reply runtime/domain.c]: Fixed.

File: runtime/domain.c (hunk 3 above)
Comment (kayceesrk) [reply runtime/domain.c]: Done.

**Comment by kayceesrk** [comment]
Thanks for the reviews!

**Pull request merged**